Issue 158  interpolation

kohr-h opened the pull request:
This PR simplifies the interpolator structures significantly by using one class for both arrays and meshgrids and specifying the variant as init argument instead. The principle is to keep the private interpolator classes simple and implement the decision logic in the callers (the interpolation operators).

Secondly, linear interpolation (#158) now works as specified in our internal discretization schemes. This means that all interpolators assume that the next "virtual" node outside the region has value zero. Hence, the value at the boundary is half the value at the last node in the region for linear interpolation. Proper boundary values are yet to be implemented.

Thirdly, interpolation schemes can now be specified per axis with the `PerAxisInterpolation` operator (#159). In fact, linear interpolation is just a special case of this operator. Nearest neighbor, however, has its own implementation for performance reasons (about 3x faster for arrays, about 2^(dim+1) times for meshes).

All new features and some old ones come with new tests.

Remaining TODOs: make new issues for
- [x] Cubic spline interpolation
- [x] Boundary values (grids with "half basis functions" at the ends), including periodic interpolation
- [x] Add per-axis support to `DiscreteLp` and adapt norm, inner product etc.
- [x] proper tessellations 

EDIT: new issues for the checklist points

adler-j commented:
> Cubic spline interpolation -> new issue?

New issue

> Boundary values (grids with "half basis functions" at the ends)

Also new issue

> Periodic interpolation

Also new issue, perhaps same as "Boundary values" since they are very related.

> Add per-axis support to DiscreteLp
 
New issue
 
> Adapt norm, inner product etc.

Same issue as above

> New issue: proper tessellations

:+1:

I'll try to give you a review of this.

adler-j commented on .pylintrc:
```diff
@@ -42,11 +42,15 @@ load-plugins=
 # and also backslash to enable :math: and not callable since it fails with inherited __call__.
 disable=redefined-builtin,
-  anomalous-backslash-in-string
+  anomalous-backslash-in-string,
```
Why these in such a branch?

adler-j commented on .pylintrc:
```diff
@@ -121,7 +125,7 @@ indent-after-paren=4
 
 # Good variable names which should always be accepted, separated by a comma
-good-names=i,j,k,ex,Run,_,a,b,x,y,z,xp,yp,zp,s,q,p,I,T,dx,dy,x1,x2
+good-names=i,j,k,ex,Run,_,a,b,x,y,z,xp,yp,zp,s,q,p,I,T,dx,dy,x1,x2,xi,yi,n,m,ax
```
Wrong branch but we'll let it slide.

kohr-h replied:
Bad old habit :-)

adler-j commented on odl/discr/discr_mappings.py:
```diff
@@ -15,19 +15,22 @@
 from future import standard_library
 standard_library.install_aliases()
-from builtins import str, zip
+from builtins import str, super, zip
```
We kindoff decided to be careful with super for performance reasons (slow as hell in the python2 backport).

kohr-h replied:
You're right, I'll have a look into its use and if it's potentially slowing down things.

kohr-h replied:
These are used only in the private classes, once per interpolator creation, i.e. once per operator evaluation.

adler-j commented on odl/discr/discr_mappings.py:
```diff
@@ -35,17 +38,22 @@
+           'GridCollocation', 'NearestInterpolation', 'LinearInterpolation',
+           'PerAxisInterpolation')
+
+SUPPORTED_INTERP_SCHEMES = ['nearest', 'linear']
```
Perhaps if this is internal it should have a `_`?

kohr-h replied:
:+1:

adler-j commented on odl/discr/discr_mappings.py:
```diff
@@ -35,17 +38,22 @@
 
     """Abstract base class for function set discretization mappings."""
 
+    # pylint: disable=abstract-method
```
Sceptical to having these in to many places. Any chance to modify this externally somehow?

kohr-h replied:
I know and agree, but I chose this over not disabling (because the warnings are nonsense) and disabling globally since warnings can actually help to find bugs. Maybe not so much warnings like `abstract-method` and `private-access`.

adler-j commented on odl/discr/discr_mappings.py:
```diff
@@ -353,14 +372,14 @@ def _call(self, x, out=None):
+        We test nearest neighbor interpolation with a non-scalar
+        data type in 2d:
+
+        >>> from __future__ import print_function
```
No way to get around this?

kohr-h replied:
Should not be needed at all, actually.

adler-j commented on odl/discr/discr_mappings.py:
```diff
@@ -467,219 +482,307 @@ def _call(self, x, out=None):
                     self.grid.coord_vectors,
-                    x.data.reshape(self.grid.shape, order=self.order))
+                    x.data.reshape(self.grid.shape, order=self.order),
+                    typ='meshgrid')
```
`typ`, dunno what i think about this name. How about `grid` or `layout`?

kohr-h replied:
Or `in_type`. I think I saw `typ` as a replacement for `type` as argument somewhere. Not so great, yeah.

adler-j commented on odl/discr/discr_mappings.py:
```diff
@@ -467,219 +482,307 @@ def _call(self, x, out=None):
-                # TODO: check if points are ordered. If not, use
-                # LinearNDInterpolator
-                interp = _LinearPointwiseInterpolator(
+                interp = _LinearInterpolator(
```
Tbh writing out `interpolator` should be done here.

kohr-h replied:
ok

adler-j commented on odl/discr/discr_mappings.py:
```diff
@@ -467,219 +482,307 @@ def _call(self, x, out=None):
+        fspace : `FunctionSpace`
+            The undiscretized (abstract) space of functions to be
+            discretized. Its field must be the same as that of data
+            space. Its `Operator.domain` must be an
```
`FunctionSpace.domain` (if it does not exist, we should probably add it)

kohr-h replied:
:+1:

adler-j commented on odl/discr/discr_mappings.py:
```diff
@@ -467,219 +482,307 @@ def _call(self, x, out=None):
+            the domain of the function set.
+        dspace : `FnBase`
+            Data space providing containers for the values of a
+            discretized object. Its size must be equal to the
```
\`FnBase.size\`

kohr-h replied:
:+1:

adler-j commented on odl/discr/discr_mappings.py:
```diff
@@ -467,219 +482,307 @@ def _call(self, x, out=None):
+            discretized object. Its size must be equal to the
+            total number of grid points. Its field must be the same
+            as that of the function space.
+        schemes : `str` or `list` of `str`
```
Is list correct, shouldnt it be sequence?

kohr-h replied:
Oh, you're right.

adler-j commented on odl/discr/discr_mappings.py:
```diff
@@ -467,219 +482,307 @@ def _call(self, x, out=None):
+            Indicates which interpolation scheme to use for which axis.
+            A single string is interpreted as a global scheme for all
+            axes.
+        nn_variants : `str` or `list` of `str`, optional
```
sequence of str

adler-j commented on odl/discr/discr_mappings.py:
```diff
@@ -467,219 +482,307 @@ def _call(self, x, out=None):
+            A single string is interpreted as a global scheme for all
+            axes.
+        nn_variants : `str` or `list` of `str`, optional
+            Which variant ('left' or 'right') to use in nearest neighbor
```
Since the use of this kw depends on the choice of `schemes` this could be rendered unused. I'd thus make it part of `kwargs`, or at the least put it last.

kohr-h replied:
That's where it should have been. Numpy does it this way: make it an item of the same kind, just at the end.

adler-j commented on odl/discr/discr_mappings.py:
```diff
@@ -467,219 +482,307 @@ def _call(self, x, out=None):
+        except TypeError:
+            schemes_ = [str(scm).lower() for scm in schemes]
+
+        if nn_variants is None:
```
Complain if used with non nearest neighbor

kohr-h replied:
:+1:

adler-j commented on odl/discr/discr_mappings.py:
```diff
@@ -467,219 +482,307 @@ def _call(self, x, out=None):
+            except TypeError:
+                variants_ = [str(var).lower() for var in nn_variants]
+
+        for i, (scm, var) in enumerate(zip(schemes, variants_)):
```
Dislike using `variants_` if `scheme` is not `nn`.

kohr-h replied:
I'll disallow it.

adler-j commented on odl/discr/discr_mappings.py:
```diff
@@ -467,219 +482,307 @@ def _call(self, x, out=None):
+
+    @property
+    def schemes(self):
+        """List of interpolation schemes, one for each axis."""
```
Why not simply return a single value if they are all the same?

kohr-h replied:
Can do.

kohr-h replied:
Second thought, not sure about this. I rather return the same type regardless of input during init. Otherwise, each time we use this value for iteration, we need to differentiate.

adler-j replied:
Ok, valid point. We should however have a repr method on this, and in that one, we can use the shortened version.

kohr-h replied:
That's true. I'm gonna add it.

adler-j commented on odl/discr/discr_mappings.py:
```diff
@@ -467,219 +482,307 @@ def _call(self, x, out=None):
+
+    @property
+    def nn_variants(self):
+        """List of nearest neighbor variants, one for each axis."""
```
Same as above.

kohr-h replied:
Yes

kohr-h replied:
See above.

adler-j commented on odl/discr/discr_mappings.py:
```diff
@@ -467,219 +482,307 @@ def _call(self, x, out=None):
+        """
+        def interpolator(arg, out=None):
+            """Interpolating function with vectorization."""
+            if is_valid_input_meshgrid(arg, self.grid.ndim):
```
Why not simply:

```python
if is_valid_input_meshgrid(arg, self.grid.ndim):
    typ = 'meshgrid'
else:
    typ = 'array'
    
interp = _PerAxisInterpolator(
    self.grid.coord_vectors,
    x.data.reshape(self.grid.shape, order=self.order),
    schemes=self.schemes, nn_variants=self.nn_variants,
    typ=typ)
```

kohr-h replied:
Hm, the code must have been more different before. Can be easily unified, right.

adler-j commented on odl/discr/discr_mappings.py:
```diff
@@ -467,219 +482,307 @@ def _call(self, x, out=None):
+        return self.range.element(interpolator, vectorized=True)
+
+
+class _Interpolator(object):
```
Shouldnt this come before the `_PerAxisInterpolator`?

kohr-h replied:
It does

adler-j commented on odl/discr/discr_mappings.py:
```diff
@@ -467,219 +482,307 @@ def _call(self, x, out=None):
+    def __init__(self, coord_vecs, values, typ):
+        """Initialize a new instance.
+
+        coord_vecs : `tuple` of `numpy.ndarray`
```
sequence

kohr-h replied:
Yep

adler-j commented on odl/discr/discr_mappings.py:
```diff
@@ -467,219 +482,307 @@ def _call(self, x, out=None):
-#            if not np.all(np.diff(p) > 0.):
-#                raise ValueError("The points in dimension {} must be strictly"
-#                                 " ascending".format(i))
         self.grid = tuple([np.asarray(p) for p in coord_vecs])
```
`self.grid = tuple(np.asarray(p) for p in coord_vecs)` should also work

kohr-h replied:
Right, I think I just copied the code.

adler-j commented on odl/discr/discr_mappings.py:
```diff
@@ -467,219 +482,307 @@ def _call(self, x, out=None):
-        shape (n, dim), where n is the number of points.
+        Parameters
+        ----------
+        x : meshgrid or `numpy.ndarray`
```
Do we have meshgrid written in the glossary?

kohr-h replied:
Good one, I'll add it.

adler-j commented on odl/discr/discr_mappings.py:
```diff
@@ -467,219 +482,307 @@ def _call(self, x, out=None):
-                             ''.format(x.shape[0], ndim))
+        if self.type == 'array':
+            # Make a (1, n) array from one with shape (n,)
+            ndim = len(self.grid)
```
Move outside of if statement and use `self.grid.ndim` instead.

adler-j commented on odl/discr/discr_mappings.py:
```diff
@@ -467,219 +482,307 @@ def _call(self, x, out=None):
+            # Make a (1, n) array from one with shape (n,)
+            ndim = len(self.grid)
+            if ndim == 1 and x.ndim == 1:
+                x = x[None, :]
```
I would use `x = x.reshape([ndim, -1])` as it communicates the goal better. Then you could also remove the two following tests since this would fail and we can rely on numpys testing instead of writing our own.

kohr-h replied:
I'll do that change.

adler-j commented on odl/discr/discr_mappings.py:
```diff
@@ -467,219 +482,307 @@ def _call(self, x, out=None):
+                                 ''.format(x.shape[0], ndim))
+            out_shape = (x.shape[1],)
+        else:
+            if len(x) != len(self.grid):
```
Use `ndim` as moved outside of the if statement above.

adler-j commented on odl/discr/discr_mappings.py:
```diff
@@ -467,219 +482,307 @@ def _call(self, x, out=None):
+                raise ValueError('number of vectors in x is {} instead of '
+                                 'the grid dimension {}.'
+                                 ''.format(len(x), len(self.grid)))
+            if len(x) == 1:
```
This is done all the time, how about we make a utility for it? (and make an issue on numpy that `broadcast` should work with a single argument.)

kohr-h replied:
Actually we have a utility, which I forgot to use. I'll look into the NumPy issue tracker if there is something about this topic (e.g. why they don't implement it).

kohr-h replied:
There's an [issue in NumPy](https://github.com/numpy/numpy/pull/6905) about this now. We'll use our utility in the meanwhile.

adler-j commented on odl/discr/discr_mappings.py:
```diff
@@ -687,42 +790,185 @@ def _evaluate(self, indices, norm_distances, out=None):
+    # means "too low", larger than or equal to 1 means "too high"
+    lo = np.where(ndist < 0)
+    hi = np.where(ndist > 1)
+    ones = np.ones_like(ndist)
```
`np.where` supports broadcasting, these shouldnt be needed.

adler-j commented on odl/discr/discr_mappings.py:
```diff
@@ -687,42 +790,185 @@ def _evaluate(self, indices, norm_distances, out=None):
+
+    # For "too high" nodes, the upper neighbor gets weight zero;
+    # "too low" gets 1.
+    if variant == 'left':
```
Not so sure about this. Numerical stability tells me that this will have some real issues and since floating point behaviour is not 100% defined, we could even get implementation dependent behavior. I cant think of many cases where we could guarantee this feature to work as people expect it to:

```python
>>> np.array([0.2 + 0.1, 0.3]) <= 0.3
array([ False, True], dtype=bool)
```

kohr-h replied:
I see the point. Maybe the situation can be rescued by changing the perspective in that case by checking for `ndist > 0.5` and change the implementation of `_evaluate`.

adler-j replied:
Possibly, and I can certainly see where this comes from mathematically, but I have a hard time seeing this working in practice since we always need to assume floats are approximate.

adler-j commented on odl/discr/discr_mappings.py:
```diff
@@ -687,42 +790,185 @@ def _evaluate(self, indices, norm_distances, out=None):
+    """Helper for nearest interpolation mimicing the linear case."""
+    # Get out-of-bounds indices from the norm_distances. Negative
+    # means "too low", larger than or equal to 1 means "too high"
+    lo = np.where(ndist < 0)
```
Why do we use `where` here?

kohr-h replied:
In an earlier version I needed the indices, too. We should be able to do without now.

adler-j commented on odl/discr/discr_mappings.py:
```diff
@@ -687,42 +790,185 @@ def _evaluate(self, indices, norm_distances, out=None):
+
+    # For "too high" nodes, the upper neighbor gets weight zero;
+    # "too low" gets 1 + yi (since yi < 0)
+    w_hi = np.array(ndist, copy=True)
```
`numpy.copy()` is more clear

kohr-h replied:
ok

adler-j commented on odl/discr/discr_mappings.py:
```diff
@@ -687,42 +790,185 @@ def _evaluate(self, indices, norm_distances, out=None):
+    return w_lo, w_hi, edge
+
+
+def _create_lists(indices, norm_distances, schemes, variants):
```
Perhaps a more specific name

kohr-h replied:
Not too great creativity here, I admit.

adler-j commented on odl/discr/discr_mappings.py:
```diff
@@ -687,42 +790,185 @@ def _evaluate(self, indices, norm_distances, out=None):
+        low_weights, high_weights, edge_indices = _create_lists(
+            indices, norm_distances, self.schemes, self.nn_variants)
+
+        # Iterate over all possible combinations of [i, i+1] for each
```
Not the most obvious logic, make sure this is tested for corner cases.

kohr-h replied:
Sure. This should be robust, however, since it's equivalent to [the original scipy inmplementation](https://github.com/scipy/scipy/blob/master/scipy/interpolate/interpolate.py#L1773-L1786).

adler-j replied:
Great!

adler-j commented on odl/discr/discr_mappings.py:
Should be omit-able.

adler-j commented on odl/discr/discr_mappings.py:
`schemes = 'linear'`

adler-j commented on odl/discr/discretization.py:
```diff
@@ -33,11 +33,8 @@
-else:
-    CudaRn = CudaCn = CudaFn = CudaNtuples = type(None)
+from odl.space.cu_ntuples import CudaNtuples, CudaFn, CudaRn
+CudaCn = type(None)  # TODO: add CudaCn to imports once it is implemented
```
I guess this is getting more and more requested, I'll look at it.

kohr-h replied:
:+1:

adler-j commented on odl/discr/discretization.py:
```diff
@@ -186,6 +183,7 @@ def element(self, inp=None):
         if inp is None:
             return self.element_type(self, self.dspace.element())
         try:
+            # pylint: disable=not-callable
```
Dislike these disables. pylint seems to fail us here, both for `element_type` and all `Operator`'s, id say disable globally.

kohr-h replied:
Ok

adler-j commented on odl/discr/discretization.py:
```diff
@@ -235,14 +233,8 @@ class RawDiscretizationVector(NtuplesBaseVector):
-                            ''.format(ntuple,
-                                      space.dspace.__class__.__name__))
+        assert isinstance(space, RawDiscretization)
+        assert isinstance(ntuple, space.dspace.element_type)
```
`assert ntuple in space.dspace`?

adler-j commented on odl/discr/discretization.py:
```diff
@@ -416,26 +408,32 @@ def one(self):
 
     def _lincomb(self, a, x1, b, x2, out):
         """Raw linear combination."""
+        # pylint: disable=protected-access
```
arghh, lets just use the public method instead then? Less error prone I guess.

kohr-h replied:
Okay, but that does not save us everywhere, e.g. in `OperatorSum`.

adler-j commented on odl/discr/discretization.py:
```diff
@@ -450,9 +448,7 @@ class DiscretizationVector(RawDiscretizationVector, FnBaseVector):
-        if not isinstance(space, Discretization):
-            raise TypeError('space {!r} not a `Discretization` '
-                            'instance.'.format(space))
+        assert isinstance(space, Discretization)
```
Using asserts in these cases is totally fine IMO, these shouldnt realy be used by users and this makes the code more simple to read.

kohr-h replied:
That was my thought, too.

adler-j commented on odl/discr/discretization.py:
```diff
@@ -450,9 +448,7 @@ class DiscretizationVector(RawDiscretizationVector, FnBaseVector):
-            raise TypeError('space {!r} not a `Discretization` '
-                            'instance.'.format(space))
+        assert isinstance(space, Discretization)
         super().__init__(space, data)
```
Perhaps `FnBaseVector.__init__` instead? Performance and lets you see which of the super-classes is called at a glance.

kohr-h replied:
Sure

adler-j commented on odl/discr/grid.py:
```diff
@@ -1165,6 +1185,7 @@ def uniform_sampling(intv_prod, num_nodes, as_midp=True):
         raise ValueError('number of nodes {} has non-positive entries.'
                          ''.format(num_nodes))
 
+    # pylint: disable=protected-access
```
Make `ideg` public instead.

kohr-h replied:
Ok

adler-j commented on odl/discr/lp_discr.py:
```diff
@@ -101,7 +96,11 @@ def __init__(self, fspace, grid, dspace, exponent=2.0, interp='nearest',
+            extension = LinearInterpolation(fspace, grid, dspace,
+                                            order=self.order)
         else:
+            # Should not happen
```
Is `raise NotImplementedError` correct then? Perhaps `assert False`?

kohr-h replied:
We could also raise `RuntimeError`, I think you did that at some place.

adler-j commented on odl/space/ntuples.py:
```diff
@@ -31,13 +31,12 @@
 from math import sqrt
 import numpy as np
-import scipy as sp
+import scipy.linalg as linalg
```
Avoid importing `a.b.c`, it harms mock and is somewhat irritating to deal with in code since it is not obvious what `linalg` is intended. Simply use `sp.linalg` instead.

kohr-h replied:
That was the first version, which failed. The problem is the following (you need to store the script as a file and run it in a new interpreter, not in Spyder):
```
>>> import scipy as sp
>>> sp.linalg.blas.get_blas_funcs('axpy')
Traceback (most recent call last):
...
AttributeError: 'module' object has no attribute 'linalg'
```
Whereas this works:
```
>>> import scipy.linalg as linalg
>>> linalg.blas.get_blas_funcs('axpy')  # ok
```
I know it's harder to mock, but the original code triggered an error when I removed some scipy imports in the `discr_mappings.py` module.

adler-j commented on test/benchmark/discr/discr_mappings_test.py:
```diff
@@ -0,0 +1,107 @@
+
+
+def initialize():
+    global shape, values, hcube, grid, space, dspace32, dspace64
```
ugh, no better way?

kohr-h replied:
Alternative would be to just not make it a function. But I thought it might be useful to have a figure on the setup time, too. Or maybe the function could return all that stuff.

adler-j commented on test/benchmark/discr/discr_mappings_test.py:
```diff
@@ -0,0 +1,107 @@
+# You should have received a copy of the GNU General Public License
+# along with ODL.  If not, see <http://www.gnu.org/licenses/>.
+
+
```
Description should be added.

kohr-h replied:
Sure.

adler-j commented on test/benchmark/discr/discr_mappings_test.py:
```diff
@@ -0,0 +1,107 @@
+    space = odl.FunctionSpace(hcube)
+    dspace32 = odl.Rn(grid.ntotal, dtype='float32')
+    dspace64 = odl.Rn(grid.ntotal, dtype='float64')
+    points = np.random.rand(4, 60 ** 4)
```
60? See above comment.

kohr-h replied:
Why not? 60 is a good number in terms of prime number decomposition. If we want an example with `ndim=5` and `ndim=6` with the same number of nodes, it's easily possible.
And, btw, this does not in any way have to do with the 100 in `shape`.

adler-j replied:
Noticed that when reading on, perhaps we can split them or add some comment here. Tests are also somewhat intended as tests and should (hopefully) be readable to outsiders (not you/me that is).

adler-j commented on test/benchmark/discr/discr_mappings_test.py:
```diff
@@ -0,0 +1,107 @@
+def initialize():
+    global shape, values, hcube, grid, space, dspace32, dspace64
+    global points, mesh
+    shape = (100, 100, 100, 100)
```
I'd write this:

```python
n = 100
ndim = 4

shape = [100] * ndim
values = np.zeros(np.prod(shape))
hcube = odl.IntervalProd([0] * ndim, [1] * ndim)
grid = odl.uniform_sampling(hcube, shape, as_midp=True)
space = odl.FunctionSpace(hcube)
dspace32 = odl.Rn(n ** ndim, dtype='float32')
dspace64 = odl.Rn(n ** ndim, dtype='float64')
points = np.random.rand(ndim, n ** ndim)
mesh = sparse_meshgrid(*([np.random.rand(n)] * ndim))
```

kohr-h replied:
Can do

adler-j commented on test/benchmark/discr/discr_mappings_test.py:
```diff
@@ -0,0 +1,107 @@
+
+
+dspace_params = [dspace32, dspace64]
+dspace_ids = [' Rn, dtype = {} '.format(spc.dtype) for spc in dspace_params]
```
`['{!r}'.format(spc) for spc in dspace_params]`

kohr-h replied:
Alright

adler-j commented on test/benchmark/discr/discr_mappings_test.py:
```diff
@@ -0,0 +1,107 @@
+
+dspace_params = [dspace32, dspace64]
+dspace_ids = [' Rn, dtype = {} '.format(spc.dtype) for spc in dspace_params]
+dspace_fixture = pytest.fixture(scope="module", ids=dspace_ids,
```
Defined and used on next row, simply use it instantly instead.

kohr-h replied:
Ok

adler-j commented on test/benchmark/discr/discr_mappings_test.py:
```diff
@@ -0,0 +1,107 @@
+
+input_params = [points, mesh]
+input_ids = [' Array input ', ' Meshgrid input ']
+input_fixture = pytest.fixture(scope="module", ids=input_ids,
```
As above

kohr-h replied:
:+1:

adler-j commented on test/benchmark/discr/discr_mappings_test.py:
```diff
@@ -0,0 +1,107 @@
+
+
+if __name__ == '__main__':
+    pytest.main(str(__file__.replace('\\', '/')) + ' -v')
```
Perhaps we could rewrite this using `sys.path`. Thats for later i presume.

kohr-h replied:
It should have a default implementation somewhere, right.

adler-j commented on test/discr/discr_mappings_test.py:
```diff
@@ -0,0 +1,392 @@
+# You should have received a copy of the GNU General Public License
+# along with ODL.  If not, see <http://www.gnu.org/licenses/>.
+
+
```
Docstring.

kohr-h replied:
Ok

adler-j commented on test/discr/discr_mappings_test.py:
```diff
@@ -0,0 +1,392 @@
+    # Testing two midpoints and the extreme values
+    pts = np.array([0.4, 0.8, 0.0, 1.0])
+    true_arr = [2, 4, 0, 4]
+    assert all_equal(function(pts), true_arr)
```
Pytest allows you to specify better error messages for these, perhaps we should look into it.

kohr-h replied:
Which would be implemented in `all_equal`, I guess?

adler-j replied:
Yeah, it was just a random note I had when reading.

adler-j commented on test/discr/discr_mappings_test.py:
```diff
@@ -0,0 +1,392 @@
+    function = interp_op(values)
+
+    # Testing two midpoints and the extreme values
+    pts = np.array([0.4, 0.8, 0.0, 1.0])
```
Does pts need to be a `np.array`?

kohr-h replied:
Currently yes. The problem is that general sequences can be interpreted both ways, i.e. is
```python
>>> x = [[1, 2],
         [3, 4]]
```
an array of two points or a meshgrid in 2D? It can be both. We need some way to distinguish, so here [I'm strict](https://github.com/odlgroup/odl/blob/issue-158__interpolation/odl/util/vectorization.py#L41-L42) and require an input array to be `ndarray`, while meshgrids are [a bit more flexible](https://github.com/odlgroup/odl/blob/issue-158__interpolation/odl/util/vectorization.py#L49-L66).

adler-j replied:
Ok im fine with this for now. Better use the strict sense until we know better.

kohr-h replied:
Right. And for the record, meshgrids are the intended default use, at least as long as we have a tensor grid to evaluate on. They're much more efficient. Arrays as input are intended for arbitrary point collections, which we don't use internally at the moment, but that may change over time.

adler-j commented on test/discr/discr_mappings_test.py:
```diff
@@ -0,0 +1,392 @@
+    # 'left' variant
+    interp_op = NearestInterpolation(space, grid, dspace, variant='left')
+    values = np.arange(5, dtype='float64')
+    function = interp_op(values)
```
Dont we have a `interp_op([1.0, 2.0, 3.0, 4.0, 5.0])` syntax?

kohr-h replied:
Should work. It follows the same rules as `Operator` in general, so any input understood by `Ntuples.element` is okay. I can use your version to improve clarity.

adler-j replied:
If `dtype='float64'` isnt necessary we could use `interp_op([1, 2, 3, 4, 5])`

kohr-h replied:
The `dtype` is only needed for `arange` since it produces an `int` array by default. The operator would convert, but maybe the test code relying on the array would have undesired results. I'll just change to the simple list.

adler-j commented on test/discr/discr_mappings_test.py:
```diff
@@ -0,0 +1,392 @@
+    # [0.1, 0.3, 0.5, 0.7, 0.9]
+
+    space = odl.FunctionSpace(intv)
+    dspace = odl.Fn(grid.ntotal, dtype='float64')
```
I propose renaming `grid.ntotal` -> `grid.size` for consistency reasons.

kohr-h replied:
Then we must rename [the current `size` method](https://github.com/odlgroup/odl/blob/issue-158__interpolation/odl/discr/grid.py#L292-L294) to something else. Maybe `extent`? In the long run, this functionality will go to a new `Tessellation` class, I guess.

adler-j replied:
Options:

* extent
* sides
* edges
* sidelen(gth)
* edgelen(gth)



Inspiration:

https://en.wikipedia.org/wiki/Square
Uses "sides" seemingly everywhere.

adler-j commented on test/discr/discr_mappings_test.py:
```diff
@@ -0,0 +1,392 @@
+    values = np.arange(8, dtype='float64')
+    function = interp_op(values)
+
+    # Evaluate at single point
```
More spacing here for readability.

adler-j commented on test/discr/discr_mappings_test.py:
```diff
@@ -0,0 +1,392 @@
+                    LinearInterpolation(space, grid, dspace, order='C'),
+                    PerAxisInterpolation(space, grid, dspace, order='C',
+                                         schemes=['linear', 'nearest'])]
+    interp_ops_f = [NearestInterpolation(space, grid, dspace, order='F',
```
Why one `_c` and one `_f`?

adler-j commented:
Some general comments, structure looks good. Since we added more abstraction, we should make sure we didnt make anything slow. Else fix and pull imo.

kohr-h commented on odl/discr/discr_mappings.py:
This is a private class, I think we don't need convenience here.

kohr-h commented on odl/discr/discr_mappings.py:
Same as above. The logic to allow for this is in the public operator, I deliberately kept this class as simple as possible.

adler-j commented on odl/discr/discr_mappings.py:
Ok that's totally fine.

kohr-h commented:
We have benchmarks on Jenkins, look there if you want to see numbers. I wrote some lines in the beginning why I think the relations are reasonable.
What I could do is add a check if our implementation is about as fast as the original Scipy one. That one could fail if we're off by a too large factor.

kohr-h commented:
All suggested changes applied. The only open question is whether we want a left and a right variant of nearest neighbor or not. The disadvantage in not having it is that there is no discretization which naturally leads to backward differences in derivative discretizations. But maybe this is not so bad, one can still explicitly use backward difference even if it's not natural for the discretization. We could simply remove the `nn_variants` option and always go for the left version.

kohr-h commented:
I added now also a speed test against Scipy for our linear interpolation. A quick check says that our version is about 20 % faster than the one in Scipy, so I added a test which fails if we are slower than Scipy.

adler-j commented:
> All suggested changes applied. The only open question is whether we want a left and a right variant of nearest neighbor or not. The disadvantage in not having it is that there is no discretization which naturally leads to backward differences in derivative discretizations. But maybe this is not so bad, one can still explicitly use backward difference even if it's not natural for the discretization. We could simply remove the nn_variants option and always go for the left version.

I certainly see where this comes from, and since we have it, I assume we can keep it, but with some disclaimers in the doc. In 98% of situations users will select the default anyway.

> I added now also a speed test against Scipy for our linear interpolation. A quick check says that our version is about 20 % faster than the one in Scipy, so I added a test which fails if we are slower than Scipy.

A 20% margin cannot hold on all platforms? Seems to tight, I'd relax it to say we are about as fast as scipy (say < 2 * scipy) to avoid un-needed fails.

adler-j commented:
Also, it seems this is the first benchmark we have (at all), I'd leave it for later and make a branch with some benchmarks so we can add them at once, then we get to consider the structure some more before we settle.

kohr-h commented:
> Also, it seems this is the first benchmark we have (at all), I'd leave it for later and make a branch with some benchmarks so we can add them at once, then we get to consider the structure some more before we settle.

OK, I'll remove the code from the branch and merge the rest. There will be a new issue branch with the benchmark then.
